### DISCUSSION TITLE
Improve file system routing for APIs

### DISCUSSION THREAD

**Discussion by rtritto** [magne4000/universal-middleware#86] [Ideas]
Some feature of [universal-autoloader](https://github.com/node-ecosystem/universal-autoloader) that can be added to handle filesystem routing of APIs:
1. Sort routes instead of [grouping](https://hono.dev/docs/api/routing#grouping) routes (same behavior and similar performance):
  https://github.com/node-ecosystem/universal-autorouter/blob/07a32680ff503b426cbd595c7b0073f6585ba2a5/src/utils.ts#L5-L11
2. Use dynamic method based on filename.
  Eg with Hono:
    ```ts
    // /api/user/(post).ts
    import type { Context } from 'hono'
    
    export default (c: Context) => {
      return c.text('Hello World!')
    }
    ```
  
    Under the hood (related [code](https://github.com/node-ecosystem/universal-autorouter/blob/07a32680ff503b426cbd595c7b0073f6585ba2a5/src/index.ts#L103-L106)):
    ```ts
    const matchedFile = filepath.match(/\/?\((.*?)\)/)
    const method = matchedFile ? matchedFile[1] as Method : 'get'  // if no method matched, use GET method
    app[method](route, importedRouteFunction)
   ```
3. Other features:
  [directory-structure](https://github.com/node-ecosystem/universal-autoloader?tab=readme-ov-file#directory-structure)

**Comment by rtritto** [comment]
What about to use a specific option to load APIs?

This will improve rendering as only Node request and not Web one (related file: [vike-handler](https://github.com/vikejs/vike-node/blob/034fb165cd9aa9d4325ddf419b4ec29ba4cd3a27/packages/vike-node/src/runtime/vike-handler.ts)).

E.g.:

```ts
import vike from 'vike-node/hono'

app.use(vike({
  apiPath: 'api'  // APIs source path
}))
```

@magne4000 @brillout what do you think?

**Comment by magne4000** [comment]
This seems out of scope of `universal-middleware`.
Vike already have its own FS routing handling, and [unrouting](https://github.com/unjs/unrouting) exists and already fulfills the purpose of being generic.
If this subject is important to you, I would suggest to participate in `unrouting` development.

**Comment by rtritto** [comment]
> Vike already have its own FS routing handling, and [unrouting](https://github.com/unjs/unrouting) exists and already fulfills the purpose of being generic.

So these features should be implemented directly in the FS routing handling in Vike?

> If this subject is important to you, I would suggest to participate in unrouting development.

My use case is to add `api` folder in a vike project and then load using the FS routing with features:
- index routes
- dynamic params with `/[<ROUTE_PARAM_NAME>].ts` syntax
- dynamic method with `/(<METHOD>).ts` syntax
- grouping (implemented with a simple sort)

Other optional (imo) features:
- wildcard `/[...id].ts` (`/*` in the route)
- wildcard with `/@[...]` prefix (already implemented with wildcard implementation)
- apply usage in frontend (eg support jsx-tsx extensions)

Source: [directory-structure](https://github.com/node-ecosystem/universal-autorouter?tab=readme-ov-file#directory-structure)

With current implementation of Vike, what is the basic way to link an API folder (eg `/api` folder) with (for example) Hono? I didn't find any example or doc reference.

PS: as another unrelated point, to support `unrouting`, I opened https://github.com/unjs/unrouting/issues/35

**Reply by magne4000** [reply]
It also is out-of-scope of Vike (at least for the moment).
One could write a dedicated `vike-*` plugin for instance to handle such use cases.

**Reply by rtritto** [reply]
Thanks for the answer.

This could be a critical point for Vike ecosystem where api routes can't be integrated in the backend server and then called by frontend.

@brillout what do you think?

**Reply by brillout** [reply]
Tbh I don't understand what you mean. In general: https://vike.dev/api-routes. Please trust us that we know what we're doing, and please be more careful and thoughtful before pinging us. As you can imagine, we don't have the capacity to teach and explain design to each individual user.

It's probably best if you reach out only if you have a *concrete* issue, instead of pinging us if you believe you've found a better design. In other words: users => complain about concrete issues, maintainer => design solutions. Only maintainers can do the design part because only the maintainers have a holistic understanding of the entire scope.

To be clear: feedback welcome. But let's keep it to that: *concrete* feedback about *concrete* issues. Thank you.

**Reply by rtritto** [reply]
> Tbh I don't understand what you mean.

The use case is the API routes without using a Telefunc or similar external tools.
With many API routes, the filesystem autoloading needs to be used. It should be integrated (maybe using `vike-node`) through a middleware and the selected web server (eg Hono).

A filesystem autoloader like [universal-autorouter](https://github.com/node-ecosystem/universal-autorouter) that already have the `viteDevServer.ssrLoadModule` implementation) can be used.

Some alternative to use `universal-autorouter`:

1. in [vike-node/runtime/vike-handler](https://github.com/vikejs/vike-node/blob/main/packages/vike-node/src/runtime/vike-handler.ts)

2. in [initializeServerEntry](https://github.com/vikejs/vike-node/blob/f80f5d857bc08dc3ce1c6107dd3e8bcb06dc668a/packages/vike-node/src/plugin/plugins/devServerPlugin.ts#L94) function. Draft:
    ```diff
    + import autoloadRoutes = 'universal-autorouter'
    
      async function initializeServerEntry(vite: ViteDevServer) {
        assert(resolvedConfig.server)
        const { index } = resolvedConfig.server.entry
        const indexResolved = await vite.pluginContainer.resolveId(index.entry)
        assert(indexResolved?.id)
        entryAbs = indexResolved.id
        vite.ssrLoadModule(entryAbs).catch(logRestartMessage)
    +   // Load APIs of /server/api folder
    +   await autoloadRoutes(app, {
    +     prefix: '/api',
    +     routesDir: '/server/api',
    +     viteDevServer: vite
    +   })
      }
    ```
3. in [vike-node/plugin/plugins/devServerPlugin](https://github.com/vikejs/vike-node/blob/main/packages/vike-node/src/plugin/plugins/devServerPlugin.ts)

**Reply by rtritto** [reply]
Opened PR https://github.com/vikejs/vike-node/pull/34

**Reply by brillout** [reply]
Response [here](https://github.com/vikejs/vike-node/issues/35#issuecomment-2552280863).
